mozilla-services/consvc-shepherd
# [SHEP-143] Unit test for Unmanaged Views

## References

[SHEP-143](https://mozilla-hub.atlassian.net/browse/SHEP-143)

## Problem Statement

Our current views are manually generated and don't have a method to enforce alignment with the models that back the views. This leads to errors when the view's SQL is updated with different fields that what was defined in the model.

## Proposed Changes

Add a new unit test that scans the database for all unmanaged models and ensures that the views that we manually create match up to the model definitions.

## Verification Steps

You can test this out in 2 ways. 
1. Edit the the CampaignSummary model without editing the view. This should fail the test
2. Edit the campaign_summary_view in the migration 0028 without updating the model. This will also fail.

Note: To test this quickly without errors add `DROP VIEW IF EXISTS ... ` to the migration file, then edit the view SQL. I've included a simple change as an example
```
    migrations.RunSQL(
            """
            DROP VIEW IF EXISTS campaign_summary_view;
            CREATE OR REPLACE VIEW campaign_summary_view AS
            SELECT
                d.id AS deal_id,
                d.id AS deal_id2,
                ...
```
               

## Visuals
Example error when the view is missing a column defined in the model
![image](https://github.com/user-attachments/assets/da40a987-4967-4fec-b1b7-7c96587a97d2)

Example error when the view doesn't match the model because it has extra columns that don't exist in the model
![image](https://github.com/user-attachments/assets/2b2227ae-5b47-47d9-8a6b-1d2cf3499a7e)



[SHEP-143]: https://mozilla-hub.atlassian.net/browse/SHEP-143?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
